client/ayon_core/pipeline/publish/lib.py

```diff
@@ -694,67 +640,91 @@ def get_custom_staging_dir_info(
+    )
 
-    Args:
-        project_name (str)
-        host_name (str)
-        product_type (str)
-        task_name (str)
-        task_type (str)
-        product_name (str)
-        project_settings(Dict[str, Any]): Prepared project settings.
-        anatomy (Dict[str, Any])
-        log (Logger) (optional)
+    if not tr_data:
+        return None, None
+
+    return tr_data["template"], tr_data["persistence"]
+
+
+def get_instance_staging_dir(instance):
+    """Unified way how staging dir is stored and created on instances.
+
+    First check if 'stagingDir' is already set in instance data.
+    In case there already is new tempdir will not be created.
 
     Returns:
-        (tuple)
-    Raises:
-        ValueError - if misconfigured template should be used
+        str: Path to staging dir
     """
-    settings = project_settings or get_project_settings(project_name)
-    custom_staging_dir_profiles = (settings["core"]
-                                           ["tools"]
-                                           ["publish"]
-                                           ["custom_staging_dir_profiles"])
-    if not custom_staging_dir_profiles:
-        return None, None
+    staging_dir = instance.data.get("stagingDir")
 
-    if not log:
-        log = Logger.get_logger("get_custom_staging_dir_info")
+    if staging_dir:
+        return staging_dir
 
-    filtering_criteria = {
-        "hosts": host_name,
-        "families": product_type,
-        "task_names": task_name,
-        "task_types": task_type,
-        "subsets": product_name
-    }
-    profile = filter_profiles(custom_staging_dir_profiles,
-                              filtering_criteria,
-                              logger=log)
+    anatomy_data = instance.data["anatomyData"]
+    template_data = copy.deepcopy(anatomy_data)
 
-    if not profile or not profile["active"]:
-        return None, None
+    product_type = instance.data["productType"]
+    product_name = instance.data["productName"]
+
+    # context data based variables
+    project_entity = instance.context.data["projectEntity"]
+    folder_entity = instance.context.data["folderEntity"]
+    task_entity = instance.context.data["taskEntity"]
+    host_name = instance.context.data["hostName"]
+    project_settings = instance.context.data["project_settings"]
+    anatomy = instance.context.data["anatomy"]
+    current_file = instance.context.data.get("currentFile")
+
+    # add current file as workfile name into formatting data
+    if current_file:
+        workfile = os.path.basename(current_file)
+        workfile_name, _ = os.path.splitext(workfile)
+        template_data["workfile_name"] = workfile_name
+
+    staging_dir_info = get_staging_dir_info(
+        host_name,
+        project_entity,
+        folder_entity,
+        task_entity,
+        product_type,
+        product_name,
+        anatomy,
+        project_settings=project_settings,
+        template_data=template_data,
+    )
```

Comment:
I would skip the part storing data to variables to just pass it to a function.
```suggestion
    # context data based variables
    context = instance.context
    current_file = context.data.get("currentFile")

    # add current file as workfile name into formatting data
    if current_file:
        workfile = os.path.basename(current_file)
        workfile_name, _ = os.path.splitext(workfile)
        template_data["workfile_name"] = workfile_name

    staging_dir_info = get_staging_dir_info(
        context.data["hostName"],
        context.data["projectEntity"],
        context.data.get("folderEntity"),
        context.data.get("taskEntity"),
        instance.data["productType"],
        instance.data["productName"],
        context.data["anatomy"],
        project_settings=context.data["project_settings"],
        template_data=template_data,
    )
```


NOTE: Why are used folder and task entities from context? They should be taken from instance, or? For example traypublisher don't have set none of them.